lib/logstorage/block.go

```diff
@@ -41,42 +41,43 @@ func (b *block) reset() {
 
 // uncompressedSizeBytes returns the total size of the original log entries stored in b.
 //
-// It is supposed that every log entry has the following format:
-//
-// 2006-01-02T15:04:05.999999999Z07:00 field1=value1 ... fieldN=valueN
-func (b *block) uncompressedSizeBytes() uint64 {
-	rowsCount := uint64(b.Len())
-
-	// Take into account timestamps
-	n := rowsCount * uint64(len(time.RFC3339Nano))
+// It uses JSON format to calculate the size as if each log entry were represented as JSON.
+func (b *block) uncompressedSizeBytes() int {
+	rowsCount := b.Len()
+	if rowsCount == 0 {
+		return 0
+	}
+
+	totalSize := (len("{}\n") + len(`"_time":""`) + len(time.RFC3339Nano)) * rowsCount
+
+	// size of constant fields (included in every row)
+	constFields := make([]Field, len(b.constColumns))
+	for i := range b.constColumns {
+		cc := &b.constColumns[i]
+		constFields[i] = Field{
+			Name:  cc.Name,
+			Value: cc.Value,
+		}
+		totalSize += estimatedJSONField(cc.Name, cc.Value) * rowsCount
+	}
 
-	// Take into account columns
-	cs := b.columns
-	for i := range cs {
-		c := &cs[i]
-		nameLen := uint64(len(c.name))
-		if nameLen == 0 {
-			nameLen = uint64(len("_msg"))
+	// add size of variable fields
+	for i := range b.columns {
+		c := &b.columns[i]
+		name := c.name
+		if name == "" {
+			name = "_msg"
 		}
+
 		for _, v := range c.values {
-			if len(v) > 0 {
-				n += nameLen + 2 + uint64(len(v))
+			if len(v) == 0 {
+				continue
```

Comment:
It would be great adding a comment explaining why we skip empty values here. The explanation is the following: VictoriaLogs data model treats empty values as non-existing values - see https://docs.victoriametrics.com/victorialogs/keyconcepts/#data-model . Empty values in a block can appear if there is at least a single non-empty value for the given column across other logs stored in this block.